false+ in IdentityBinaryExpression

## What version of Error Prone are you using?
2.2.0

Warning message is:

C:\Users\noel\Documents\GitHub\h2database-noel\h2\src\main\org\h2\table\TableLink.java:124: error: [IdentityBinaryExpression] A binary expression where both operands are the same is usually incorrect; the value of this expression is equivalent to `rs.next()`.
        if (rs.next() && rs.next()) {
                      ^
    (see http://errorprone.info/bugpattern/IdentityBinaryExpression)
1 error

Code looks like:
        ResultSet rs = meta.getTables(null, originalSchema, originalTable, null);
        if (rs.next() && rs.next()) {
            throw DbException.get(ErrorCode.SCHEMA_NAME_MUST_MATCH, originalTable);
        }

Probably ResultSet::next should be treated like Iterator next method

--- a/check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java
+++ b/check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java
@@ -558,6 +558,7 @@ public class ASTHelpers {
    *
    * @param annotationClass the binary class name of the annotation (e.g.
    *     "javax.annotation.Nullable", or "some.package.OuterClassName$InnerClassName")
+   * @return true if the symbol is annotated with given type.
    */
   public static boolean hasAnnotation(Symbol sym, String annotationClass, VisitorState state) {
     Name annotationName = state.getName(annotationClass);
@@ -603,12 +604,23 @@ public class ASTHelpers {
   /**
    * Check for the presence of an annotation, considering annotation inheritance.
    *
-   * @return the annotation of given type on the tree's symbol, or null.
+   * @param annotationClass the binary class name of the annotation (e.g.
+   *     "javax.annotation.Nullable", or "some.package.OuterClassName$InnerClassName")
+   * @return true if the tree is annotated with given type.
+   */
+  public static boolean hasAnnotation(Tree tree, String annotationClass, VisitorState state) {
+    Symbol sym = getDeclaredSymbol(tree);
+    return hasAnnotation(sym, annotationClass, state);
+  }
+
+  /**
+   * Check for the presence of an annotation, considering annotation inheritance.
+   *
+   * @return true if the tree is annotated with given type.
    */
   public static boolean hasAnnotation(
       Tree tree, Class<? extends Annotation> annotationClass, VisitorState state) {
-    Symbol sym = getDeclaredSymbol(tree);
-    return hasAnnotation(sym, annotationClass.getName(), state);
+    return hasAnnotation(tree, annotationClass.getName(), state);
   }
 
   /**
@@ -626,6 +638,18 @@ public class ASTHelpers {
       }
     }
     return false;
+  }
+
+  /**
+   * Check for the presence of an annotation with a specific simple name directly on this symbol.
+   * Does *not* consider annotation inheritance.
+   *
+   * @param tree the tree to check for the presence of the annotation
+   * @param simpleName the simple name of the annotation to look for, e.g. "Nullable" or
+   *     "CheckReturnValue"
+   */
+  public static boolean hasDirectAnnotationWithSimpleName(Tree tree, String simpleName) {
+    return hasDirectAnnotationWithSimpleName(getDeclaredSymbol(tree), simpleName);
   }
 
   /**

--- a/core/src/main/java/com/google/errorprone/bugpatterns/collectionincompatibletype/TruthIncompatibleType.java
+++ b/core/src/main/java/com/google/errorprone/bugpatterns/collectionincompatibletype/TruthIncompatibleType.java
@@ -1,0 +1,146 @@
+/*
+ * Copyright 2018 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns.collectionincompatibletype;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.bugpatterns.EqualsIncompatibleType;
+import com.google.errorprone.bugpatterns.EqualsIncompatibleType.TypeCompatibilityReport;
+import com.google.errorprone.bugpatterns.collectionincompatibletype.AbstractCollectionIncompatibleTypeMatcher.MatchResult;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.matchers.method.MethodMatchers.ParameterMatcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.google.errorprone.util.Signatures;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.ParenthesizedTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.TypeCastTree;
+import com.sun.source.util.SimpleTreeVisitor;
+import com.sun.tools.javac.code.Type;
+import javax.annotation.Nullable;
+
+/** @author cushon@google.com (Liam Miller-Cushon) */
+@BugPattern(
+  name = "TruthIncompatibleType",
+  summary = "Argument is not compatible with the subject's type.",
+  severity = WARNING
+)
+public class TruthIncompatibleType extends BugChecker implements MethodInvocationTreeMatcher {
+
+  private static final AbstractCollectionIncompatibleTypeMatcher MATCHER =
+      new AbstractCollectionIncompatibleTypeMatcher() {
+
+        // TODO(cushon): expand to other assertThat methods to handle e.g. ListSubject
+        private final ParameterMatcher assertThatObject =
+            MethodMatchers.staticMethod()
+                .onClass("com.google.common.truth.Truth")
+                .named("assertThat")
+                .withParameters("java.lang.Object");
+
+        public final Matcher<ExpressionTree> isEqualTo =
+            MethodMatchers.instanceMethod()
+                // TODO(cpovirk): Extend to subclasses, blacklisting any with unusual behavior.
+                .onExactClass("com.google.common.truth.Subject")
+                .named("isEqualTo")
+                .withParameters("java.lang.Object");
+
+        @Override
+        Matcher<ExpressionTree> methodMatcher() {
+          return isEqualTo;
+        }
+
+        @Nullable
+        @Override
+        Type extractSourceType(MethodInvocationTree tree, VisitorState state) {
+          return ASTHelpers.getType(getOnlyElement(tree.getArguments()));
+        }
+
+        @Nullable
+        @Override
+        ExpressionTree extractSourceTree(MethodInvocationTree tree, VisitorState state) {
+          return getOnlyElement(tree.getArguments());
+        }
+
+        @Nullable
+        @Override
+        Type extractTargetType(MethodInvocationTree tree, VisitorState state) {
+          for (ExpressionTree curr = tree;
+              curr instanceof MethodInvocationTree;
+              curr = ASTHelpers.getReceiver(curr)) {
+            if (assertThatObject.matches(curr, state)) {
+              return ASTHelpers.getType(
+                  Iterables.getOnlyElement(((MethodInvocationTree) curr).getArguments())
+                      .accept(
+                          new SimpleTreeVisitor<Tree, Void>() {
+                            @Override
+                            protected Tree defaultAction(Tree node, Void unused) {
+                              return node;
+                            }
+
+                            @Override
+                            public Tree visitTypeCast(TypeCastTree node, Void unused) {
+                              return node.getExpression().accept(this, null);
+                            }
+
+                            @Override
+                            public Tree visitParenthesized(ParenthesizedTree node, Void unused) {
+                              return node.getExpression().accept(this, null);
+                            }
+                          },
+                          null));
+            }
+          }
+          return null;
+        }
+      };
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    MatchResult result = MATCHER.matches(tree, state);
+    if (result == null) {
+      return Description.NO_MATCH;
+    }
+    TypeCompatibilityReport compatibilityReport =
+        EqualsIncompatibleType.compatibilityOfTypes(
+            result.targetType(), result.sourceType(), state);
+    if (compatibilityReport.compatible()) {
+      return Description.NO_MATCH;
+    }
+    String sourceType = Signatures.prettyType(result.sourceType());
+    String targetType = Signatures.prettyType(result.targetType());
+    if (sourceType.equals(targetType)) {
+      sourceType = result.sourceType().toString();
+      targetType = result.targetType().toString();
+    }
+    Description.Builder description = buildDescription(tree);
+    description.setMessage(
+        String.format(
+            "Argument '%s' should not be passed to this method; its type %s is not compatible "
+                + "with subject type %s",
+            result.sourceTree(), sourceType, targetType));
+    return description.build();
+  }
+}

--- a/core/src/main/java/com/google/errorprone/refaster/ControlFlowVisitor.java
+++ b/core/src/main/java/com/google/errorprone/refaster/ControlFlowVisitor.java
@@ -51,14 +51,14 @@ import javax.lang.model.element.Name;
  *
  * @author lowasser@google.com (Louis Wasserman)
  */
-class ControlFlowVisitor extends SimpleTreeVisitor<Result, BreakContext> {
+public class ControlFlowVisitor extends SimpleTreeVisitor<Result, BreakContext> {
   public static final ControlFlowVisitor INSTANCE = new ControlFlowVisitor();
 
   /**
    * The state of whether a sequence of statements may return, break out of the visited statements,
    * or neither.
    */
-  enum Result {
+  public enum Result {
     NEVER_EXITS {
       @Override
       Result or(Result other) {

--- a/core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java
+++ b/core/src/main/java/com/google/errorprone/scanner/BuiltInCheckerSuppliers.java
@@ -248,6 +248,7 @@ import com.google.errorprone.bugpatterns.argumentselectiondefects.AutoValueConst
 import com.google.errorprone.bugpatterns.collectionincompatibletype.CollectionIncompatibleType;
 import com.google.errorprone.bugpatterns.collectionincompatibletype.CompatibleWithMisuse;
 import com.google.errorprone.bugpatterns.collectionincompatibletype.IncompatibleArgumentType;
+import com.google.errorprone.bugpatterns.collectionincompatibletype.TruthIncompatibleType;
 import com.google.errorprone.bugpatterns.formatstring.FormatString;
 import com.google.errorprone.bugpatterns.formatstring.FormatStringAnnotationChecker;
 import com.google.errorprone.bugpatterns.inject.AssistedInjectAndInjectOnConstructors;
@@ -524,6 +525,7 @@ public class BuiltInCheckerSuppliers {
           StringSplitter.class,
           StaticGuardedByInstance.class,
           StreamResourceLeak.class,
+          TruthIncompatibleType.class,
           SynchronizeOnNonFinalField.class,
           ThreadJoinLoop.class,
           ThreadLocalUsage.class,

--- a/core/src/test/java/com/google/errorprone/bugpatterns/collectionincompatibletype/TruthIncompatibleTypeTest.java
+++ b/core/src/test/java/com/google/errorprone/bugpatterns/collectionincompatibletype/TruthIncompatibleTypeTest.java
@@ -1,0 +1,65 @@
+/*
+ * Copyright 2018 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns.collectionincompatibletype;
+
+import com.google.errorprone.CompilationTestHelper;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** {@link TruthIncompatibleType}Test */
+@RunWith(JUnit4.class)
+public class TruthIncompatibleTypeTest {
+
+  private final CompilationTestHelper compilationHelper =
+      CompilationTestHelper.newInstance(TruthIncompatibleType.class, getClass());
+
+  @Test
+  public void positive() {
+    compilationHelper
+        .addSourceLines(
+            "Test.java",
+            "import static com.google.common.truth.Truth.assertThat;",
+            "public class Test {",
+            "  static final class A {}",
+            "  static final class B {}",
+            "  public void f(A a, B b) {",
+            "    // BUG: Diagnostic contains:",
+            "    assertThat(a).isEqualTo(b);",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void negative() {
+    compilationHelper
+        .addSourceLines(
+            "Test.java",
+            "import static com.google.common.truth.Truth.assertThat;",
+            "public class Test {",
+            "  static final class A {}",
+            "  static final class B {}",
+            "  public void f(A a, B b) {",
+            "    assertThat(a).isEqualTo(a);",
+            "    assertThat(b).isEqualTo(b);",
+            "    assertThat(\"a\").isEqualTo(\"b\");",
+            "  }",
+            "}")
+        .doTest();
+  }
+}

--- a/docs/bugpattern/EqualsIncompatibleType.md
+++ b/docs/bugpattern/EqualsIncompatibleType.md
@@ -75,7 +75,7 @@ coordinates), this generally means that the equals method will only return true
 if the other object has the exact same class, and each of the components is
 equal to the corresponding component in the other object. Here, there are
 numerous tools in the Java ecosystem to generate the appropriate `equals` and
-`hashCode` method implementations, including Guava's [AutoValue][av].
+`hashCode` method implementations, including [AutoValue][av].
 
 Another pattern often seen is to declare a common supertype with a defined
 `equals` method (like `List`, which defines equality by having equal elements in

--- a/docs/bugpattern/IdentityBinaryExpression.md
+++ b/docs/bugpattern/IdentityBinaryExpression.md
@@ -1,14 +1,27 @@
-`a && a`, `a || a`, `a & a`, or `a | a`
-:   equivalent to `a`
+Using the same expressions as both arguments to the following binary expressions
+is usually a mistake:
 
-`a <= a`, `a >= a`, or `a == a`
-:   always `true`
+*   `a && a`, `a || a`, `a & a`, or `a | a` is equivalent to `a`
+*   `a <= a`, `a >= a`, or `a == a` is always `true`
+*   `a < a`, `a > a`, `a != a`, or `a ^ a` is always `false`
+*   `a / a` is always `1`
+*   `a % a` or `a - a` is always `0`
 
-`a < a`, `a > a`, `a != a`, or `a ^ a`
-:   always `false`
+If the expression has side-effects, consider refactoring one of the expressions
+with side effects into a local. For example, prefer this:
 
-`a / a`
-:   always `1`
+```.java {.good}
+// check twice, just to be sure
+boolean isTrue = foo.isTrue();
+if (isTrue && foo.isTrue()) {
+  // ...
+}
+```
 
-`a % a` or `a - a`
-:   always `0`
+to this:
+
+```.java {.bad}
+if (foo.isTrue() && foo.isTrue()) {
+  // ...
+}
+```

